Python/Product/Uap/Debugger/PythonRemoteDebugEvents.cs

```diff
@@ -0,0 +1,118 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Xml.Linq;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Debugger;
+using Microsoft.VisualStudio.Debugger.ComponentInterfaces;
+using Microsoft.VisualStudio.Debugger.Exceptions;
+using Microsoft.VisualStudio.Debugger.Interop;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.PythonTools.Uap.Debugger {
+    internal class PythonRemoteDebugEvents : IVsDebuggerEvents, IDebugEventCallback2, IDkmExceptionTriggerHitNotification {
+        private static readonly Lazy<PythonRemoteDebugEvents> instance = new Lazy<PythonRemoteDebugEvents>();
+
+        // This exception code is a Win32 exception that is expected to be thrown by the debug client to trigger this code flow
+        // i.e. whatever remote process is launching the debug server, should throw and catch this exception code before starting
+        // remote Python debug server to have automatic attach work
+        internal const uint RemoteDebugStartExceptionCode = 0xEDCBA987;
+
+        public const string RemoteDebugExceptionId = "E7DD0845-FB1A-4A45-8192-44953C0ACC51";
+        public static readonly Guid RemoteDebugExceptionGuid = new Guid(RemoteDebugExceptionId);
+
+        public static PythonRemoteDebugEvents Instance {
+            get { return instance.Value; }
+        }
+
+        public Func<System.Threading.Tasks.Task> AttachRemoteProcessFunction { get; set; }
+
+        public string AttachRemoteDebugXml { get; set; }
+
+        public int Event(IDebugEngine2 pEngine, IDebugProcess2 pProcess, IDebugProgram2 pProgram, IDebugThread2 pThread, IDebugEvent2 pEvent, ref Guid riidEvent, uint dwAttrib) {
+            try {
+                if (riidEvent == typeof(IDebugProgramCreateEvent2).GUID) {
+                    Guid processId;
+
+                    // A program was created and attached
+                    if (pProcess != null) {
+                        if (VSConstants.S_OK == pProcess.GetProcessId(out processId)) {
+                            DkmProcess dkmProcess = DkmProcess.FindProcess(processId);
+
+                            if (dkmProcess != null) {
+                                var debugTrigger = DkmExceptionCodeTrigger.Create(DkmExceptionProcessingStage.Thrown, null, DkmExceptionCategory.Win32, RemoteDebugStartExceptionCode);
+
+                                // Try to add exception trigger for when a remote debugger server is started for Python
+                                dkmProcess.AddExceptionTrigger(RemoteDebugExceptionGuid, debugTrigger);
+                            }
+                        }
+                    }
+                }
+
+                return VSConstants.S_OK;
+            } finally {
+                if (pEngine != null && Marshal.IsComObject(pEngine)) {
+                    Marshal.ReleaseComObject(pEngine);
```

Comment:
It's dangerous in general because you don't know whether the RCW that you get is shared with someone else or not, especially if you don't control its creation. We don't do it in our debugger code, and I think it's best to avoid it here as well.